### PR TITLE
Save new user data

### LINKED ISSUE #13: Backend - API: Save new user data
## User story

**Who:** **As a** developer
**What:** **I want to** be able to create a new user register in the Database
**Why:** **so that I can** save a new user data and then retrieve it later.

## Acceptance criteria

Create the PUT endpoint to save  a new user:  `/api/users`

image

 
Check the wireframes here: https://www.figma.com/file/yo4QT2EcPo83HAMI3ydMxb/HYF-Wireframes?node-id=4%3A1324
and the DB Diagram https://dbdiagram.io/d/616ae287940c4c4eec9881e3

## Implementation details

Make sure that you implement the endpoint into Swagger https://github.com/HackYourFuture-CPH/fp-class19#working-with-api-documentation---swagger

### PR DESCRIPTION
**Pull request by Divyajg**
# Description

Created an endpoint to save the new user data.

Fixes # (13)

# How to test?

Pull the Branch, 
run: npm run dev ,
go to: [http://localhost:3000/api/documentation], 
You will see: Users below the Modules,
got to: Postman and make a post request with body and you should see `{
    "successful": true
}`
go to database: you should see the new user created in the users table.

# Checklist

- [✔ ] I have performed a self-review of my own code
- [ ✔] I have followed the name conventions for CSS Classnames and filenames, Components names and filenames, Style filenames, if you are in doubt check the project README.MD and here https://github.com/HackYourFuture-CPH/curriculum/blob/master/review/review-checklist.md
- [ ✔] I have commented on my code, particularly in hard-to-understand areas, if your code was simple enough mark the box anyway
- [ ✔] I have made corresponding changes to the documentation, if your code was simple enough mark the box anyway
- [✔ ] This PR is ready to be merged and not breaking any other functionality

newuser

### CODE REVIEW

**Review by FlorentinaPetica — APPROVED** [review]
Hey Divya, I've tested your pull request. It works as it should. I was able to see the documentation for swagger and I was also able to make a post request for a new users. It is ready for merge. Great job 👍

File: src/server/api/routes/api-router.js [hunk 1]
```diff
@@ -30,5 +31,6 @@ router.use('/documentation', swaggerUi.serve, swaggerUi.setup(swaggerDocument));
 
 // Application routes
 router.use('/modules', modulesRouter);
+router.use('/users', usersRouter);
```
Comment (senner007) [inline src/server/api/routes/api-router.js]: this must be "/user" and not "/users"

File: src/server/api/routes/users.router.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+ *      5XX:
+ *        description: Unexpected error.
+ */
+router.post('/', (req, res) => {
+  usersController
+    .createUser(req.body)
+    .then((result) => res.json(result))
+    .catch((error) => {
+      console.log(error);
+
```
Comment (senner007) [inline src/server/api/routes/users.router.js]: Here you return 400, but your swagger docs says nothing of this

**Review by senner007 — CHANGES_REQUESTED** [review]

File: src/server/api/routes/users.router.js (hunk 2 above)
Comment (senner007) [reply src/server/api/routes/users.router.js]: Remember to get into the habit of creating prs when the work is done everything is verified/tested

**Review by saloumeh-67 — COMMENTED** [review]
Thanks @Divyajg ,it works on my machine ;).

File: src/server/api/routes/users.router.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+ *    tags:
+ *    - Users
+ *    summary: Create a user
+ *    description:
+ *      Will create a user.
+ *    produces: application/json
+ *    parameters:
+ *      - in: body
+ *        name: user
+ *        description: The module to create.
```
Comment (senner007) [inline src/server/api/routes/users.router.js]: This need fixing also. It must say 'The use to create'

File: src/server/api/routes/users.router.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ *            - full_name
+ *            - email
+ *            - address
+ *            - zipcode
+ *            - city
+ *            - country
+ *            - created_at
+
+ *          properties:
+ *            full_name:
```
Comment (senner007) [inline src/server/api/routes/users.router.js]: Here you should insert user props instead of "string".
Remember we test in swagger, not Postman

**Review by senner007 — CHANGES_REQUESTED** [review]

**Review by suvarna-ratna — APPROVED** [review]
Working as it should..

**Comment by Divyajg** [comment]
@senner007 Please review the changes.

File: src/server/api/controllers/users.controller.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+const knex = require('../../config/db');
+const moment = require('moment-timezone');
+
+const createUser = async (body) => {
```
Comment (senner007) [inline src/server/api/controllers/users.controller.js:4]: Hi. You will want to add additional validation here. Remember a city can not be number, for instance.
Validation should happen in its own function. I suggest just creating an issue for now

File: src/server/api/routes/users.router.js [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+ *         type: string
+ *       address:
+ *         type: string
+ *       zipcode:
+ *         type: integer
+ *       city:
+ *         type: string
+ *       country:
+ *         type: string
+ */
```
Comment (senner007) [inline src/server/api/routes/users.router.js:52]: Here I made some corrections. This is not optimal but we can refactor later

File: src/server/api/controllers/users.controller.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+  await knex('users').insert({
```
Comment (senner007) [inline src/server/api/controllers/users.controller.js:5]: Hi. You will want to add additional validation here. Remember a city can not be number, for instance.
Validation should happen in its own function. I suggest just creating an issue for now

**Review by senner007 — APPROVED** [review]
👍
Hi. look at the changes I made and verify yourself. Add additional validation later.

**Pull request merged**